packages/react/src/ReactServerContext.js

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import {
+  REACT_PROVIDER_TYPE,
+  REACT_SERVER_CONTEXT_TYPE,
+} from 'shared/ReactSymbols';
+
+import type {
+  ReactServerContext,
+  ServerContextJSONValue,
+} from 'shared/ReactTypes';
+
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+
+import {enableServerContext} from 'shared/ReactFeatureFlags';
+
+const globalServerContextRegistry =
+  ReactSharedInternals.globalServerContextRegistry;
+
+let DEFAULT_PLACEHOLDER;
+if (enableServerContext) {
+  DEFAULT_PLACEHOLDER = globalServerContextRegistry.__defaultValue;
+}
+
+export function createServerContext<T: ServerContextJSONValue>(
+  globalName: string,
+  defaultValue: T,
+): ReactServerContext<T> {
+  if (!enableServerContext) {
+    throw new Error('Not implemented.');
+  }
+  if (!globalServerContextRegistry[globalName]) {
+    globalServerContextRegistry[globalName] = _createServerContext(
+      globalName,
+      defaultValue,
+    );
+  }
+  const context = globalServerContextRegistry[globalName];
+  if (!context._definitionLoaded) {
+    context._definitionLoaded = true;
+    context._defaultValue = defaultValue;
+  } else {
+    throw new Error(`ServerContext: ${globalName} already defined`);
+  }
+  return context;
+}
+
+function _createServerContext<T: ServerContextJSONValue>(
+  globalName: string,
+  defaultValue: T,
+): ReactServerContext<T> {
+  const context: ReactServerContext<T> = {
+    $$typeof: REACT_SERVER_CONTEXT_TYPE,
+    // As a workaround to support multiple concurrent renderers, we categorize
+    // some renderers as primary and others as secondary. We only expect
+    // there to be two concurrent renderers at most: React Native (primary) and
+    // Fabric (secondary); React DOM (primary) and React ART (secondary).
+    // Secondary renderers store their context values on separate fields.
+    __currentValue: defaultValue,
+    __currentValue2: defaultValue,
+
+    get _currentValue() {
```

Comment:
We typically don't use getters because they're usually pretty slow to install and to consume.

However, what is worse is that since this will be accessed by readContext that's shared with other uses of regular Context, that callsite will be deopted and so this risks slowing down all client side Contexts when it's used.